deepsubtree.go

```diff
@@ -0,0 +1,271 @@
+	// and set their keys correctly
+	for _, node := range nodes {
+		pnode, _ := dst.ndb.GetNode(node.hash)
+		if pnode.leftNode != nil {
+			pnode.key = pnode.leftNode.getHighestKey()
+		}
+
+		if pnode.rightNode != nil {
+			pnode.key = pnode.rightNode.getLowestKey()
+		}
+	}
+	if dst.root == nil {
+		rootNode, rootErr := dst.ndb.GetNode(rootHash)
+		if rootErr != nil {
+			return fmt.Errorf("could not set root of deep subtree: %w", rootErr)
+		}
+		dst.root = rootNode
+	}
+
+	return nil
+}
+
+// Set sets a key in the working tree with the given value.
+// Assumption: Node with given key already exists and is a leaf node.
+func (dst *DeepSubTree) Set(key []byte, value []byte) (updated bool, err error) {
+	if value == nil {
+		return updated, fmt.Errorf("attempt to store nil value at key '%s'", key)
+	}
+
+	dst.root, updated, err = dst.recursiveSet(dst.root, key, value)
+	dst.root.hash = nil
+	dst.root._hash()
+	return updated, err
+}
+
+// Helper method for set to traverse and find the node with given key
+// recursively.
+func (dst *DeepSubTree) recursiveSet(node *Node, key []byte, value []byte) (
+	newSelf *Node, updated bool, err error,
+) {
+	version := dst.version + 1
+
+	if node.isLeaf() {
+		switch bytes.Compare(key, node.key) {
+		case -1:
+			return &Node{
+				key:           node.key,
+				subtreeHeight: 1,
+				size:          2,
+				leftNode:      NewNode(key, value, version),
+				rightNode:     node,
+				version:       version,
+			}, false, nil
+		case 1:
+			return &Node{
+				key:           key,
+				subtreeHeight: 1,
+				size:          2,
+				leftNode:      node,
+				rightNode:     NewNode(key, value, version),
+				version:       version,
+			}, false, nil
+		default:
+			return NewNode(key, value, version), true, nil
+		}
+	} else {
+		node.version = version
+		leftNode, rightNode := node.leftNode, node.rightNode
+		if leftNode == nil && rightNode == nil {
+			return nil, false, fmt.Errorf("inner node must have at least one child node set")
+		}
+		compare := bytes.Compare(key, node.key)
+		if (leftNode != nil && compare < 0) || rightNode == nil {
+			node.leftNode, updated, err = dst.recursiveSet(leftNode, key, value)
+			if err != nil {
+				return nil, updated, err
+			}
+			node.leftNode.hash = nil
+			node.leftNode._hash()
+			node.leftHash = node.leftNode.hash
+		} else if (rightNode != nil && compare >= 0) || leftNode == nil {
+			node.rightNode, updated, err = dst.recursiveSet(rightNode, key, value)
+			if err != nil {
+				return nil, updated, err
+			}
+			node.rightNode.hash = nil
+			node.rightNode._hash()
+			node.rightHash = node.rightNode.hash
+		} else {
+			return nil, false, fmt.Errorf("inner node does not have key set correctly")
+		}
+		return node, updated, nil
+	}
+}
+
+// Prints a Deep Subtree recursively.
+func (dst *DeepSubTree) printNodeDeepSubtree(node *Node, indent int) error {
+	indentPrefix := ""
+	for i := 0; i < indent; i++ {
+		indentPrefix += "    "
+	}
```

Comment:
https://go.dev/play/p/ahRjg9Lf84A
```suggestion
	indentPrefix := strings.Repeat("    ", indent)
```

Reply:
I'd like to keep it consistent with this implementation: https://github.com/celestiaorg/iavl/blob/56e223198068a06c3fc1bad141b2444965ee5929/util.go#L18

Added a comment referencing it instead.